segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java

```diff
@@ -555,575 +378,700 @@ private boolean isStorageSystemSegment(SegmentMetadata segmentMetadata) {
-                if (segmentMetadata.getOwnerEpoch() < this.epoch) {
-                    log.debug("{} openRead - Segment needs ownership change. segment={}.", logPrefix, segmentMetadata.getName());
-                    // In case of a failover, length recorded in metadata will be lagging behind its actual length in the storage.
-                    // This can happen with lazy commits that were still not committed at the time of failover.
-                    claimOwnership(txn, segmentMetadata);
-                }
-                val retValue = SegmentStorageHandle.readHandle(streamSegmentName);
-                LoggerHelpers.traceLeave(log, "openRead", traceId, retValue);
-                return retValue;
+        public CompletableFuture<Integer> call() {
+            traceId = LoggerHelpers.traceEnter(log, "read", handle, offset, length);
+            timer = new Timer();
+
+            // Validate preconditions.
+            checkPreconditions();
+            streamSegmentName = handle.getSegmentName();
+            return tryWith(chunkedSegmentStorage.metadataStore.beginTransaction(streamSegmentName),
+                    txn -> txn.get(streamSegmentName)
+                            .thenComposeAsync(storageMetadata -> {
+                                segmentMetadata = (SegmentMetadata) storageMetadata;
+
+                                // Validate preconditions.
+                                checkState();
+
+                                if (length == 0) {
+                                    return CompletableFuture.completedFuture(0);
+                                }
+
+                                return findChunkForOffset(txn)
+                                        .thenComposeAsync(v -> {
+                                            // Now read.
+                                            return readData(txn);
+                                        }, chunkedSegmentStorage.executor)
+                                        .thenApplyAsync(v -> {
+                                            logEnd();
+                                            return totalBytesRead;
+                                        }, chunkedSegmentStorage.executor);
+                            }, chunkedSegmentStorage.executor),
+                    chunkedSegmentStorage.executor);
+        }
+
+        private void logEnd() {
+            Duration elapsed = timer.getElapsed();
+            log.debug("{} read - segment={}, offset={}, bytesRead={}, latency={}.", chunkedSegmentStorage.logPrefix, handle.getSegmentName(), offset, totalBytesRead, elapsed.toMillis());
+            LoggerHelpers.traceLeave(log, "read", traceId, handle, offset, totalBytesRead);
+        }
+
+        private CompletableFuture<Void> readData(MetadataTransaction txn) {
+            return Futures.loop(
+                    () -> bytesRemaining > 0 && null != currentChunkName,
```

Comment:
This is not thread safe. Here (and in other classes within this big class) you use a number of non-final fields which you frequently modify using different threads. While even though you do it in sequence, due to thread local caching of memory addresses, such modifications may not be immediately visible to other threads, and hence even for subsequent iterations of this loop.

Reply:
These methods are called on a single object one after another - so there is no possibility of **concurrent modifications**.  But there may be visibility problems. Making the mutable fields volatile and other fields final solves the visibility problem.

Reply:
Also now I have split big bad class into smaller classes.